foreman/data_refinery_foreman/surveyor/array_express.py

```diff
@@ -40,13 +40,19 @@ def determine_pipeline(self,
 
     def group_batches(self) -> List[List[Batch]]:
         """Groups batches based on the download URL of their only File"""
-        groups = []
-        download_urls = {file.download_url for file in [batch.files[0] for batch in self.batches]}
-
-        for url in download_urls:
-            groups.append([batch for batch in self.batches if batch.files[0].download_url == url])
+        # Builds a mapping of each unique download_url to a list of
+        # Batches whose first File's download_url matches.
+        download_url_mapping = {}
+        for batch in self.batches:
+            download_url = batch.files[0].download_url
+            if download_url in download_url_mapping:
+                download_url_mapping[download_url].append(batch)
+            else:
+                download_url_mapping[download_url] = [batch]
 
-        return groups
+        # The values of the mapping we built are the groups the
+        # batches should be grouped into.
+        return list(download_url_mapping.values())
```

Comment:
Isn't `download_url_mapping.values()` already a list?

Reply:
No, in fact they are [dictionary view objects](https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects). I believe they are iterable, but cannot be accessed via indices:
```python
>>> d = {"a": 1}
>>> v = d.values()
>>> v[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'dict_values' object does not support indexing
```

Reply:
You are right. This seems to be a new feature in Python 3. I forgot it.